### PR TITLE
Export TIMM image models to onnx

### PR DESCRIPTION
**Pull request by liangfu**
*Issue #, if available:*

*Description of changes:*
1. This PR exports the TIMM image models to onnx, and it's been tested on onnxruntime for checking correctness.
2. Downgrades onnx to 1.12.x (since https://github.com/autogluon/autogluon/pull/2567), due to conflict version range of protobuf against tensorboardx:
  a. onnx 1.13.x requires protobuf >=3.12.2,<=3.20.1, (onnx 1.12.x requires protobuf >=3.12.2,<=3.20.1)
  b. tensorboardx 2.5.1 requires protobuf >=3.8.0,<=3.20.1

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by liangfu** [comment]
cc @FANGAreNotGnu @zhiqiangdon @tonyhoo

File: multimodal/src/autogluon/multimodal/models/utils.py [hunk 1]
```diff
@@ -684,3 +684,19 @@ def update_mmdet_config(key, value, config):
             for subsubconfig in subconfig:
                 if isinstance(subsubconfig, dict):
                     update_mmdet_config(key, value, subsubconfig)
+
+
+def run_model(model: nn.Module, batch: dict):
+    if type(model).__module__.endswith("timm_image"):
```
Comment (tonyhoo) [inline multimodal/src/autogluon/multimodal/models/utils.py:690]: Wondering if all backbones in TIMM will be ONNX compatible or just the ones we use as of now

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 2]
```diff
@@ -3147,3 +3155,28 @@ def __init__(self, **kwargs):
...
+
+
+class MultiModalOnnxPredictor(MultiModalPredictor):
+    def __init__(self, model):
+        import onnxruntime as ort
+
+        self.sess = ort.InferenceSession(model.SerializeToString(), providers=["CPUExecutionProvider"])
+
+    def predict(self, data: Union[pd.DataFrame, dict, list, str]):
+        raise NotImplementedError()
```
Comment (tonyhoo) [inline multimodal/src/autogluon/multimodal/predictor.py:3167]: Wondering why we left this unimplemented?

Comment (liangfu) [reply multimodal/src/autogluon/multimodal/predictor.py:3167]: This depends on problem_type / postprocessing, which isn't available within the converted onnx model.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 3]
```diff
@@ -2771,10 +2771,19 @@ def export_onnx(
...
+        self._model.to(device).eval()
+        batch = move_to_device(batch, device=device)
+
+        if type(self._model).__module__.endswith("timm_image"):
+            input_vec = tuple([batch[k] for k in self._model.input_keys])
+        else:
+            input_vec = batch
         torch.onnx.export(
             self._model.eval(),
-            batch,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/predictor.py:2776]: Can we still use the dict format as model input? It seems that @FANGAreNotGnu made it work when export hf_text models.

Comment (liangfu) [reply multimodal/src/autogluon/multimodal/predictor.py:2776]: The reason we are planning to remove away from dict input comes with two folds: 1) there are [limitations](https://pytorch.org/docs/stable/onnx.html#limitations) in exporting onnx, 2) there are lots of benefits if we keep model inputs / outputs clear, see https://github.com/huggingface/transformers/issues/16059

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:2776]: The dict input design is to flexibly support various modalities, backbones, and fusions in the AutoMM. I think changing the design may cause issues for some cases. For example, it seems difficult to list arguments of a fusion model since it may fuse any combinations of image, text, and tabular backbones.

File: multimodal/src/autogluon/multimodal/utils/inference.py [hunk 4]
```diff
@@ -146,6 +147,22 @@ def infer_batch(
     return output[model.prefix]
 
 
+def run_model(model: nn.Module, batch: dict):
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/inference.py:150]: Duplicate function? I seem this function also exists in `models/utils.py`.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 3 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:2776]: For most functions in AutoMM, we have used the type hints. The dict design can make each model more modular. A batch may contain inputs for several backbones and a model only needs to extract necessary information from it. This is especially useful in various fusion settings.

File: multimodal/src/autogluon/multimodal/models/utils.py (hunk 1 above)
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/utils.py:690]: Users may customize the model instance name to `timm_image_123` or `timm_image_abc`.

Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/utils.py:690]: This captures name of the module, which is indicated by `timm_image.py` file. As long as we don't change the name of the py file, we are completely compatible with current design. This has been tested with multiple test cases with `timm_image_abc` as part of batch keys.

File: multimodal/src/autogluon/multimodal/utils/inference.py (hunk 4 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/inference.py:150]: Nice catch! Yes, it is the one in `utils.py` that works. Import the funtion here would cause circular dependencies in fusion models.

Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/inference.py]: removed the duplicated function.

File: multimodal/src/autogluon/multimodal/models/timm_image.py [hunk 5]
```diff
@@ -87,21 +91,35 @@ def image_feature_dim(self):
 
     def forward(
         self,
-        batch: dict,
+        images: torch.FloatTensor,
+        image_valid_num: torch.Tensor,
+        image_column_names: Optional[List[str]] = None,
+        image_column_values: Optional[List[torch.Tensor]] = None,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/timm_image.py:97]: How about changing the name `image_column_values `  to `image_column_indexes`? It stores the start and end indexes of each column's images in `images`.

File: multimodal/src/autogluon/multimodal/models/utils.py (hunk 1 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/models/utils.py:690]: What do you think if using `if isinstance(model, TimmAutoModelForImagePrediction)`?

File: multimodal/src/autogluon/multimodal/models/timm_image.py (hunk 5 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/timm_image.py]: Renamed to `image_column_indices`.

File: multimodal/src/autogluon/multimodal/models/utils.py (hunk 1 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/utils.py]: Changed to `if isinstance(model, TimmAutoModelForImagePrediction)`

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 3 above)
Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/predictor.py:2746]: If the dict does not work, does it make sense for us to switch to dataclasses or namedtuple @liangfu @zhiqiangdon ?

Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/predictor.py:2746]: I think using namedtuple is feasible. Similar to https://github.com/huggingface/transformers/blob/45da7cec5aa1b1bf1031af9caa9085e95e262e11/examples/research_projects/bertabs/run_summarization.py#L28

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:2746]: We can try namedtuple, but not sure whether it is compatible with onnx.

**Review by zhiqiangdon — APPROVED** [review]
LGTM. Thanks for fixing onnx export! Remember resolving the conflicts with the master branch.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 3 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/predictor.py:2523]: namedtuple is now experimentally working as inputs for tracing, but not yet ready to be part of `collate` function. will try integrate it into `collate` function in a follow-up PR.

**Comment by theolundqvist** [comment]
Hi, how would I go ahead and use this feature. The output from the onnx model is not the same as when running it through autogluon. In my case (binary classification using timm) the onnx model outputs one 768 dimensional vector and one 2 dimensional vector for each image in the batch. The 2d vector is always expit(logits) = 0.5+-0.03, while the same images through autogluon receive a predicted score of 0.13-0.95, something is wrong with the compiled model.

Please add documentation or tests showing that it works.

Should I open a new issue?
